source/blog/2017-07-12-unprivileged-containers-with-bwrap-oci-and-bubblewrap.md

```diff
@@ -0,0 +1,187 @@
+
+The uid/gid defined in the OCI config.json file (i.e. the uid/gid used
+to exec the sandboxed program) must be  mapped to the uid/gid of the
+user in the parent namespace.
+Any other uid/gid in the range [0, AVAILABLE_UIDS] is mapped to the
+additional ids defined for the user in the parent namespace, as
+described above.
+
+At this point we have all we need to run systemd as a non privileged
+user.
+
+In the bwrap-oci repository there is a [demo](https://github.com/projectatomic/bwrap-oci/blob/master/demos/run-systemd/run_demo.sh)
+on how to setup a container, for simplicity I’ll skip the steps
+required to build it here.
+
+Run a container with bwrap-oci is extremely easy once you have the
+rootfs with all the files required by the container and the
+config.json,
+
+I am not root:
+
+```
+$ id -u
+1000
+```
+
+And I do not have any current capabilities:
+
+```
+$ grep ^Cap /proc/self/status
+CapInh:    0000000000000000
+CapPrm:    0000000000000000
+CapEff:    0000000000000000
+CapBnd:    0000003fffffffff
+CapAmb:    0000000000000000
+```
+
+In order to run a container I am required to have a json file
+describing the application to run and a rootfs directory structure
+containing the application:
+
+```
+$ ls -l
+-rw-rw-r--.  1 gscrivano gscrivano   5741 Jul  7 11:53 config.json
+drwxrwxr-x. 18 gscrivano gscrivano   4096 Jul  6 12:26 rootfs
+```
+
+Now I use bwrap-oci to run the container:
+
+```
+$ bwrap-oci run systemd-demo
+systemd 231 running in system mode. (+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN)
+Detected virtualization container-other.
+Detected architecture x86-64.
+
+Welcome to Fedora 25 (Twenty Five)!
+
+....
+```
+
+Thus, "systemd --system" running as an unprivileged init container!
+
+From another shell, we can get a list of the containers known by bwrap-oci:
+
+```
+$ bwrap-oci list
+NAME                          PID       STATUS    BUNDLE
+systemd-demo               7412      running   /home/gscrivano/systemd-container
+```
+
+Systemd halts the machine on receiving the signal `SIGRTMIN+3`.  I use
+directly the value of `SIGRTMIN+3` until “bwrap-oci kill” understands
+signal names (you can also use directly kill -SIGRTMIN+3 $PID):
+
+```
+$ bwrap-oci kill systemd-demo 37
+```
+
+In the first shell we can see that systemd container is stopped now.
+
+
+# Future directions:
+
+Improve the integration of bwrap-oci tool into atomic and get more
+containers that work now with ‘atomic install --system’ to work in the
+same way with ‘atomic install --user’ using bwrap-oci instead of runC
+to be executed.  User containers will use “systemd --user” instead of
+the system systemd as we do with system containers to manage their
+life cycle.
+The etcd container already works with bwrap-oci/bwrap and you should
+be able to fully run it with “atomic install --user” and “systemctl
+--user”.
```

Comment:
If you could do this quickly, it would be terrific to have etcd as a 2nd example of this feature.  I think more users will relate to it than will to systemd-demo.